fdbclient/include/fdbclient/TenantEntryCache.actor.h

```diff
@@ -102,16 +105,104 @@ class TenantEntryCache : public ReferenceCounted<TenantEntryCache<T>>, NonCopyab
 		return tenantList.results;
 	}
 
+	ACTOR static Future<Void> refreshCacheById(int64_t tenantId,
+	                                           TenantEntryCache<T>* cache,
+	                                           TenantEntryCacheRefreshReason reason) {
+		TraceEvent(SevDebug, "TenantEntryCacheIDRefreshStart", cache->id()).detail("Reason", static_cast<int>(reason));
+		state Reference<ReadYourWritesTransaction> tr = cache->getDatabase()->createTransaction();
+		loop {
+			try {
+				tr->setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+				tr->setOption(FDBTransactionOptions::READ_LOCK_AWARE);
+				state Optional<TenantName> name = wait(TenantMetadata::tenantIdIndex().get(tr, tenantId));
+				if (name.present()) {
+					Optional<TenantMapEntry> entry = wait(TenantMetadata::tenantMap().get(tr, name.get()));
+					if (entry.present()) {
+						cache->put(std::make_pair(name.get(), entry.get()));
+					}
+				}
+				updateCacheRefreshMetrics(cache, reason);
```

Comment:
nit:
1. Please move `updateCacheRefreshMetrics` only when update is done.
2. Maybe add an `else` clause if `name.present()` is not `true`; not sure if that is even a valid case...

Reply:
It's a valid case when we ask for a tenant that doesn't exist, so essentially we try to do a point refresh but find nothing even after the refresh.

Reply:
Is this referring to `invalid tenant` lookup semantics? If yes, can we do better by tracking the `maxValidTenant` known to the system and not even do database lookup? 

The only case such a lookup is possible is (hypothetically): tenantId existed once, got deleted and now either tenant with same `name or id` got recreated (hence a valid lookup)

Reply:
I did move the `updateCacheRefreshMetrics` though

Reply:
> Is this referring to invalid tenant lookup semantics?

Yes this would be the case, i guess we could do a maxValidTenant, but it would only be useful in the case you mentioned where we delete a tenant and try to look for that tenant (since we can only avoid the db if we id < maxTenantId). Not sure how often we'll run into that case so wondering if there is a point in supporting this use case until such time? I could put a TODO here in the meantime

Reply:
Getting a maxTenantId would require use to keep track of the max tenant each time a tenant was created. It seems like alot of work for this specific use case

Reply:
> Getting a maxTenantId would require use to keep track of the max tenant each time a tenant was created.

The usecase of concern for me is the following: for `TenantMode::disabled` usecase, this might translate to db-key lookup for every map access. I think it would be better to update TenantMap such that lookup is O(1) if cluster TenantMode::Disabled? Otherwise callers would need to do this check (multiple places) and failure to do this would translate to degraded performance. 

Once we handle `TenantMode::Disabled` scenario, the only scenario where `invalid-tenant` cache-misses could happen is `TenantMode::Optional`, we can add TODO for now and handle it later. 

For completeness: 
`TenantMetadaSpecification` has `lastTenantId` as a `KeyBackedProperty` defined, why do think setting a watch on it is lot of work? Again, am not proposing to do this as part of this PR and can be done as a separate PR for sure. 

Thanks

Reply:
>  I think it would be better to update TenantMap such that lookup is O(1) if cluster TenantMode::Disabled
I agree this would be helpful but currently the issue is there are two ways we can get the tenant mode; either get it from the DatabaseContext or get it from the DatabaseConfiguration. There are some instances in which the `DatabaseContext.clientDBInfo.tenantMode` may not be fully up to date (i.e it may show tenants disabled for a brief period of time before updating when its actually required). If we use the DatabaseConfiguration (source of truth), this requires us to read system keys which is not an O(1) operation. Letting clients decide to use the tenant cache or not might make more sense in this case

Reply:
Discussed offline, utilized both the tenantMode and lastTenant id